image_buffer/image_buffer.go

```diff
@@ -17,36 +17,46 @@ type ImageBuffer struct {
 	mu             sync.Mutex
 	ringBuffer     []CachedData
 	toSend         []CachedData
+	captureFrom    time.Time
 	captureTill    time.Time
 	lastCached     CachedData
-	windowSeconds  int
+	windowSecondsBefore  int
+	windowSecondsAfter 	 int
 	imageFrequency float64
 	maxImages      int
 }
 
-func NewImageBuffer(windowSeconds int, imageFrequency float64) *ImageBuffer {
+func NewImageBuffer(windowSeconds int, imageFrequency float64, windowSecondsBefore int, windowSecondsAfter int) *ImageBuffer {
 	// Calculate the maximum number of images to keep in the ring buffer
 	// Keep images for 2 * windowSeconds (before and after trigger)
-	maxImages := int(2 * float64(windowSeconds) * imageFrequency)
+	var maxImages int
+	if windowSeconds > 0 {
+		maxImages = int(2 * float64(windowSeconds) * imageFrequency)
+		windowSecondsBefore = windowSeconds
+		windowSecondsAfter = windowSeconds
+	} else if windowSecondsBefore > 0 || windowSecondsAfter > 0 {
+		maxImages = int(float64(windowSecondsBefore + windowSecondsAfter) * imageFrequency)
+	}
 	return &ImageBuffer{
 		ringBuffer:     []CachedData{},
 		toSend:         []CachedData{},
-		windowSeconds:  windowSeconds,
+		windowSecondsBefore:  windowSecondsBefore,
+		windowSecondsAfter: windowSecondsAfter,
 		imageFrequency: imageFrequency,
 		maxImages:      maxImages,
 	}
 }
 
-func (ib *ImageBuffer) windowDuration() time.Duration {
-	return time.Second * time.Duration(ib.windowSeconds)
-}
-
 func (ib *ImageBuffer) MarkShouldSend(now time.Time) {
 	ib.mu.Lock()
 	defer ib.mu.Unlock()
 
-	ib.captureTill = now.Add(ib.windowDuration())
+	// Add images from the ring buffer that are within the window
+	beforeTimeBoundary := time.Second * time.Duration(ib.windowSecondsBefore)
+	afterTimeBoundary := time.Second * time.Duration(ib.windowSecondsAfter)
 
+	ib.captureFrom = now.Add(-beforeTimeBoundary)
+	ib.captureTill = now.Add(afterTimeBoundary)
```

Comment:
This needs careful thought. Suppose we get multiple consecutive matching images: we want to keep the earliest `ib.captureFrom`, but the latest `ib.captureTill`. We only want to update `ib.captureFrom` when we transition from not capturing images to capturing them at all. I think it's something like this:
```go
newCaptureFrom := now.Add(-beforeTimeBoundary)
newCaptureTill := now.Add(afterTimeBoundary)
if ib.captureTill.Before(now) {
    // We're about to start capturing images! Update the old captureFrom.
    ib.captureFrom = newCaptureFrom
} // Otherwise, we've been capturing this whole time and should keep the old captureFrom.
ib.captureTill = newCaptureTill
```